src/smt/smt_engine.cpp

```diff
@@ -546,7 +546,13 @@ class SmtEnginePrivate : public NodeManagerListener {
 
   /** TODO: whether certain preprocess steps are necessary */
   //bool d_needsExpandDefs;
-
+  
+  //------------------------------- expression names
+  /** mapping from expressions to names
+  * This is a user-context stack
+  */
+  std::stack< std::map<Expr, std::string> > d_exprNames;
```

Comment:
I think that using a `CDHashMap` would really be cleaner. For that, we'd have to:
- Remove `getExpressionNames()` (since `CDHashMap` is private)
- Add `hasNamedExpression()` that checks whether an expression is in `d_exprNames` and `getExpressionName()` that gets the name of the expression (that should be enough to modify the [SMT2 printer](https://github.com/CVC4/CVC4/blob/master/src/printer/smt2/smt2_printer.cpp))
- Remove the `names` parameter from `UnsatCore::toStream` and `Printer:: toStream` (`UnsatCore` already has a pointer to `SmtEngine`, so it should not need `names`) 

What do you think?

Reply:
My impression was that the interface for UnsatCore is public so we should minimize changes to it.  See https://github.com/CVC4/CVC4/blob/master/src/proof/unsat_core.h#L35
That's why I hadn't changed the interface.
I think its a question of whether an API user wants to explicitly use a toStream version that takes the std::map< Expr, std::string >& names, in theory this may be a useful feature in the API.

Reply:
Thinking about this more, the first thing is to work out the expected behavior of the UnsatCore functions in proof/unsat_core.h  where there are two print functions:

(1) toStream( out ) 
(2) toStream( out, names )

Currently (1) prints the unsat core with no names, and (2) prints with (externally provided) names only.  Neither consult the SmtEngine pointer in the UnsatCore object.  Should this change?

Reply:
Users of the API were not able to "officially" name expressions with the current code (they had to keep track of the names themselves). If they kept track of the names themselves, they probably also iterated through the `Expr`s in the unsat core and did custom printing. It seems unlikely that they relied on an undocumented method. IMHO, it seems reasonable to streamline this API with the introduction of `setExpressionName()`.

As far as I could tell, (2) is only used for the SMT2 printer, other printers are just doing the same thing for both versions: https://github.com/CVC4/CVC4/blob/master/src/printer/printer.cpp#L86-L97

This makes (2) a bit hack-ish, so I'd be in favor of removing variant (2). The `toStream` methods do not need to consult the `SmtEngine` pointer, we can add the `hasNamedExpression()`/`getExpressionName()` to the UnsatCore class (which just redirect the calls to the SmtEngine). This would seem fairly clean to me: The UnsatCore "knows" about which expressions are named.

Reply:
I agree, I'll commit these changes shortly.